0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+ the current 1.51 rustc says
+ "the usage of `mac!` is _likely_ invalid in item context"
+ (emphasis mine),
+ which could be strengthened if all macro arms capture `$:value`.)
+
+For a given capture `$name:value`,
+the captured expression is evaluated a single time upon entry into the macro expansion,
+whether `$name` is mentioned in the macro expansion zero, one, or any number times.
+Every expansion of `$name` within the macro expansion refers to the name of the temporary where the captured expression was evaluated.
+If more than one `$:value` capture is present,
+they are evaluated from left to right.
+The intent is that this has identical semantics to that of a function argument capture.
+
+A compiler MAY implement this by expanding to a `match` expression,
+similar to the following:
+
+```rust
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:value /* other captures */ ) => {
+        /* macro body */
+    };
+    /* other arms */
+}
+// "desugars" to
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:expr /* other captures */ ) => {
+        match $name {
+            name => {
+                /* macro body, $name replaced with name (hygienically) */
+            }
+        }
+    };
+    /* other arms */
+}
+```
+
+but the compiler is expected to also handle the case where `$:value` is inside of a macro repetition,
+which cannot be directly implemented by just a desugaring of the `macro_rules!` invocation.
+
+To illustrate (one of) the differences in lifetimes, consider temporary lifetime extension:
+
+```rust
+macro_rules! mlet {
+    ( $e:expr ) => {{
+        let e = $e;
+        e
+    }};
+}
+
+macro_rules! mmatch {
+    ( $e:expr ) => {
+        match $e {
+            e => e,
+        }
+    };
+}
+
+fn mfn<T>(e: T) -> T {
+    e
+}
+
+struct Wrap(&'static str);
+impl Drop for Wrap {
+    fn drop(&mut self) {}
+}
+
+fn a() { let _a = mlet!  (&Wrap("a")); } // borrowck error
+fn b() { let _b = mmatch!(&Wrap("b")); } // compiles
+fn c() { let _c = mfn    (&Wrap("c")); } // compiles
+```
+
+plus there are other differences with drop timing of temporaries,
+especially with macros used in statement position rather than expression position.
+Using `:value` gives the known behavior of a function argument,
+rather than "however you used it."
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+For one, `$:value` is another thing that has to be learned to write effective `macro_rules!` macros.
+However, it replaces the `match` trick, so the author believes this comes out nuetral.
+
+Of course, it expands the language, where a code solution is sufficient, if not elegant.
+
+Additionally,
+this would be the first `macro_rules!` macro where a capture isn't plain token (tree) substitution.
+If the `match` trick is actually fully equivalent to function argument semantics,
+(the author believes this to be true but hasn't exhaustively shown this,)
```

Comment:
Spurious comma.

Reply:
A parenthetical placed at the end of a clause requiring a comma is the bane of my writing (because my voice likes using them so much, but I don't know the exact rules for where the comma is supposed to go). Also this may be one of those things that differs between American English and British English, like the placement of a comma at the end of a quote (that does not originally end with a comma, and the comma is part of the outer sentence).